### ISSUE TITLE
NoClassDefFoundError: com/google/android/collect/Maps

### ISSUE THREAD

**Issue by QuadFlask** [robolectric/robolectric#1420]
When I run tests, it throw NoClassDefFoundError 

```
java.lang.NoClassDefFoundError: com/google/android/collect/Maps
    at org.robolectric.shadows.ShadowMediaMetadataRetriever.<clinit>(ShadowMediaMetadataRetriever.java:21)
    at org.robolectric.RobolectricBase.reset(RobolectricBase.java:1811)
    at org.robolectric.Robolectric.reset(Robolectric.java:1372)
    at org.robolectric.internal.ParallelUniverse.resetStaticState(ParallelUniverse.java:37)
    at org.robolectric.RobolectricTestRunner$2.evaluate(RobolectricTestRunner.java:207)
    at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
    at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
    at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
    at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
    at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
    at org.robolectric.RobolectricTestRunner$1.evaluate(RobolectricTestRunner.java:158)
    at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
    at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.runTestClass(JUnitTestClassExecuter.java:86)
    at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.execute(JUnitTestClassExecuter.java:49)
    at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassProcessor.processTestClass(JUnitTestClassProcessor.java:69)
    at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:48)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
    at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
    at org.gradle.messaging.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
    at org.gradle.messaging.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
    at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
    at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:105)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
    at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
    at org.gradle.messaging.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:360)
    at org.gradle.internal.concurrent.DefaultExecutorFactory$StoppableExecutorImpl$1.run(DefaultExecutorFactory.java:64)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.ClassNotFoundException: com.google.android.collect.Maps
    at org.robolectric.bytecode.AsmInstrumentingClassLoader.getByteCode(AsmInstrumentingClassLoader.java:172)
    at org.robolectric.bytecode.AsmInstrumentingClassLoader.findClass(AsmInstrumentingClassLoader.java:126)
    at org.robolectric.bytecode.AsmInstrumentingClassLoader.loadClass(AsmInstrumentingClassLoader.java:98)
    ... 40 more
```

My androidstudio is fine but my friend one throw this error. (gradle clean, build works fine)

**Comment by stoefln** [comment]
Same here. @erd why did you close this issue?

**Comment by erd** [comment]
#1438 was merged, which removed references to com.google.android.collect.Maps. There's nothing else to do here. The issue above is related to the gradle plugin, which already has an open issue.

**Comment by vexdev** [comment]
@erd #1438 did not fix the issue.
We spent 3 days debugging it on any version > 2.4 of Robolectric and we finally found that Robolectric sometime is not able to load `android-all-4.3_r2-robolectric-0.jar` which contains com.google.android.collect.Maps but is also needed in many other ways by robolectric, so #1438 does not solve this, it just hides the real problem.

Now talking of 2.4 source code:
In our case this happened because CachedDependencyResolver was getting the path to this jar from a file that is stored serialized in a common directory of the system, in our case /tmp/robolectric/<number>, and if robolectric found this cache, it wont ask Maven for the proper directory for current user. Unfortunately we run robolectric with a different user and the Cache was pointing to its jar which was not accessible from the first user, so the tests of the first user stopped working.

The workaround is to clean each time your tmp directory (rm -rf /tmp/*) if you run this on a C.I. system like we did.

The proper solution needs to be applied to robolectric so that it does not always rely on this cache because currently (2.4) Robolectric basically does not support multiple users on the same machine.
